Private cached properties are included in BaseModel.__iter__

### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Attributes whose name has a leading underscore are not treated as fields by Pydantic, and are not included in the model schema. However this doesn't seem to be the case for fields injected by `cached_property`.

### Example Code

```Python
from functools import cached_property

from pydantic import BaseModel


class Model(BaseModel):
    foo: int

    @cached_property
    def _foo(self):
        return -self.foo


m = Model(foo=1)
assert list(k for k, v in m) == ["foo"]

m._foo
keys = list(k for k, v in m)
assert keys == ["foo"], keys
```

Raises `AssertionError: ['foo', '_foo']`


### Python, Pydantic & OS Version

```Text
pydantic version: 2.3.0
        pydantic-core version: 2.6.3
          pydantic-core build: profile=release pgo=false
               python version: 3.10.11 | packaged by conda-forge | (main, May 10 2023, 19:01:19) [Clang 14.0.6 ]
                     platform: macOS-13.2.1-arm64-arm-64bit
     optional deps. installed: ['email-validator', 'typing-extensions']
```

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -7,7 +7,12 @@ import sys
 import typing
 from copy import copy
 from dataclasses import Field as DataclassField
-from functools import cached_property
+
+try:
+    from functools import cached_property  # type: ignore
+except ImportError:
+    # python 3.7
+    cached_property = None
 from typing import Any, ClassVar
 from warnings import warn
 
@@ -992,14 +997,14 @@ def computed_field(__func: PropertyT) -> PropertyT:
     ...
 
 
-def _wrapped_property_is_private(property_: cached_property | property) -> bool:
+def _wrapped_property_is_private(property_: cached_property | property) -> bool:  # type: ignore
     """Returns true if provided property is private, False otherwise."""
     wrapped_name: str = ''
 
     if isinstance(property_, property):
         wrapped_name = getattr(property_.fget, '__name__', '')
-    elif isinstance(property_, cached_property):
-        wrapped_name = getattr(property_.func, '__name__', '')
+    elif isinstance(property_, cached_property):  # type: ignore
+        wrapped_name = getattr(property_.func, '__name__', '')  # type: ignore
 
     return wrapped_name.startswith('_') and not wrapped_name.startswith('__')
 

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -2,12 +2,12 @@
 from __future__ import annotations as _annotations
 
 import dataclasses
-import functools
 import inspect
 import sys
 import typing
 from copy import copy
 from dataclasses import Field as DataclassField
+from functools import cached_property
 from typing import Any, ClassVar
 from warnings import warn
 
@@ -992,13 +992,13 @@ def computed_field(__func: PropertyT) -> PropertyT:
     ...
 
 
-def _wrapped_property_is_private(property_: functools.cached_property | property) -> bool:
+def _wrapped_property_is_private(property_: cached_property | property) -> bool:
     """Returns true if provided property is private, False otherwise."""
     wrapped_name: str = ''
 
     if isinstance(property_, property):
         wrapped_name = getattr(property_.fget, '__name__', '')
-    elif isinstance(property_, functools.cached_property):
+    elif isinstance(property_, cached_property):
         wrapped_name = getattr(property_.func, '__name__', '')
 
     return wrapped_name.startswith('_') and not wrapped_name.startswith('__')
@@ -1105,6 +1105,31 @@ def computed_field(
         #> ValueError("you can't override a field with a computed field")
     ```
 
+    Private properties decorated with `@computed_field` have `repr=False` by default.
+
+    ```py
+    from functools import cached_property
+
+    from pydantic import BaseModel, computed_field
+
+    class Model(BaseModel):
+        foo: int
+
+        @computed_field
+        @cached_property
+        def _private_cached_property(self) -> int:
+            return -self.foo
+
+        @computed_field
+        @property
+        def _private_property(self) -> int:
+            return -self.foo
+
+    m = Model(foo=1)
+    print(repr(m))
+    #> M(foo=1)
+    ```
+
     Args:
         __f: the function to wrap.
         alias: alias to use when serializing this computed field, only used when `by_alias=True`
@@ -1112,7 +1137,8 @@ def computed_field(
         title: Title to used when including this computed field in JSON Schema, currently unused waiting for #4697
         description: Description to used when including this computed field in JSON Schema, defaults to the functions
             docstring, currently unused waiting for #4697
-        repr: whether to include this computed field in model repr
+        repr: whether to include this computed field in model repr.
+            Default is `False` for private properties and `True` for public properties.
         return_type: optional return for serialization logic to expect when serializing to JSON, if included
             this must be correct, otherwise a `TypeError` is raised.
             If you don't include a return type Any is used, which does runtime introspection to handle arbitrary

--- a/pydantic/fields.py
+++ b/pydantic/fields.py
@@ -2,6 +2,7 @@
 from __future__ import annotations as _annotations
 
 import dataclasses
+import functools
 import inspect
 import sys
 import typing
@@ -991,6 +992,18 @@ def computed_field(__func: PropertyT) -> PropertyT:
     ...
 
 
+def _wrapped_property_is_private(property_: functools.cached_property | property) -> bool:
+    """Returns true if provided property is private, False otherwise."""
+    wrapped_name: str = ''
+
+    if isinstance(property_, property):
+        wrapped_name = getattr(property_.fget, '__name__', '')
+    elif isinstance(property_, functools.cached_property):
+        wrapped_name = getattr(property_.func, '__name__', '')
+
+    return wrapped_name.startswith('_') and not wrapped_name.startswith('__')
+
+
 def computed_field(
     __f: PropertyT | None = None,
     *,
@@ -998,7 +1011,7 @@ def computed_field(
     alias_priority: int | None = None,
     title: str | None = None,
     description: str | None = None,
-    repr: bool = True,
+    repr: bool | None = None,
     return_type: Any = PydanticUndefined,
 ) -> PropertyT | typing.Callable[[PropertyT], PropertyT]:
     """Decorator to include `property` and `cached_property` when serializing models or dataclasses.
@@ -1118,7 +1131,13 @@ def computed_field(
         # if the function isn't already decorated with `@property` (or another descriptor), then we wrap it now
         f = _decorators.ensure_property(f)
         alias_priority = (alias_priority or 2) if alias is not None else None
-        dec_info = ComputedFieldInfo(f, return_type, alias, alias_priority, title, description, repr)
+
+        if repr is None:
+            repr_: bool = False if _wrapped_property_is_private(property_=f) else True
+        else:
+            repr_ = repr
+
+        dec_info = ComputedFieldInfo(f, return_type, alias, alias_priority, title, description, repr_)
         return _decorators.PydanticDescriptorProxy(f, dec_info)
 
     if __f is None:

--- a/pydantic/main.py
+++ b/pydantic/main.py
@@ -887,7 +887,7 @@ class BaseModel(metaclass=_model_construction.ModelMetaclass):
 
     def __iter__(self) -> TupleGenerator:
         """So `dict(model)` works."""
-        yield from self.__dict__.items()
+        yield from [(k, v) for (k, v) in self.__dict__.items() if not k.startswith('_')]
         extra = self.__pydantic_extra__
         if extra:
             yield from extra.items()

--- a/tests/test_computed_fields.py
+++ b/tests/test_computed_fields.py
@@ -402,7 +402,7 @@ def test_private_computed_field():
     class MyModel(BaseModel):
         x: int
 
-        @computed_field
+        @computed_field(repr=True)
         def _double(self) -> int:
             return self.x * 2
 

--- a/tests/test_private_attributes.py
+++ b/tests/test_private_attributes.py
@@ -395,7 +395,8 @@ class BaseConfig(BaseModel):
     assert module.BaseConfig._FIELD_UPDATE_STRATEGY == {}
 
 
-def test_private_properties_not_included_in_iter() -> None:
+@pytest.mark.skipif(not hasattr(functools, 'cached_property'), reason='cached_property is not available')
+def test_private_properties_not_included_in_iter_cached_property() -> None:
     class Model(BaseModel):
         foo: int
 
@@ -405,17 +406,25 @@ def test_private_properties_not_included_in_iter() -> None:
             return -self.foo
 
     m = Model(foo=1)
-    assert '_foo' not in list(k for k, v in m)
+    assert '_foo' not in list(k for k, _ in m)
 
 
-def test_private_properties_not_included_in_repr_by_default() -> None:
+def test_private_properties_not_included_in_iter_property() -> None:
     class Model(BaseModel):
         foo: int
 
         @computed_field
         @functools.cached_property
-        def _private_cached_property(self) -> int:
+        def _foo(self) -> int:
             return -self.foo
+
+    m = Model(foo=1)
+    assert '_foo' not in list(k for k, _ in m)
+
+
+def test_private_properties_not_included_in_repr_by_default_property() -> None:
+    class Model(BaseModel):
+        foo: int
 
         @computed_field
         @property
@@ -424,5 +433,19 @@ def test_private_properties_not_included_in_repr_by_default() -> None:
 
     m = Model(foo=1)
     m_repr = repr(m)
-    assert '_private_cached_property' not in m_repr
     assert '_private_property' not in m_repr
+
+
+@pytest.mark.skipif(not hasattr(functools, 'cached_property'), reason='cached_property is not available')
+def test_private_properties_not_included_in_repr_by_default_cached_property() -> None:
+    class Model(BaseModel):
+        foo: int
+
+        @computed_field
+        @functools.cached_property
+        def _private_cached_property(self) -> int:
+            return -self.foo
+
+    m = Model(foo=1)
+    m_repr = repr(m)
+    assert '_private_cached_property' not in m_repr

--- a/tests/test_private_attributes.py
+++ b/tests/test_private_attributes.py
@@ -414,7 +414,7 @@ def test_private_properties_not_included_in_iter_property() -> None:
         foo: int
 
         @computed_field
-        @functools.cached_property
+        @property
         def _foo(self) -> int:
             return -self.foo
 

--- a/tests/test_private_attributes.py
+++ b/tests/test_private_attributes.py
@@ -4,7 +4,7 @@ from typing import ClassVar, Generic, TypeVar
 import pytest
 from pydantic_core import PydanticUndefined
 
-from pydantic import BaseModel, ConfigDict, PrivateAttr
+from pydantic import BaseModel, ConfigDict, PrivateAttr, computed_field
 
 
 def test_private_attribute():
@@ -393,3 +393,36 @@ class BaseConfig(BaseModel):
     )
 
     assert module.BaseConfig._FIELD_UPDATE_STRATEGY == {}
+
+
+def test_private_properties_not_included_in_iter() -> None:
+    class Model(BaseModel):
+        foo: int
+
+        @computed_field
+        @functools.cached_property
+        def _foo(self) -> int:
+            return -self.foo
+
+    m = Model(foo=1)
+    assert '_foo' not in list(k for k, v in m)
+
+
+def test_private_properties_not_included_in_repr_by_default() -> None:
+    class Model(BaseModel):
+        foo: int
+
+        @computed_field
+        @functools.cached_property
+        def _private_cached_property(self) -> int:
+            return -self.foo
+
+        @computed_field
+        @property
+        def _private_property(self) -> int:
+            return -self.foo
+
+    m = Model(foo=1)
+    m_repr = repr(m)
+    assert '_private_cached_property' not in m_repr
+    assert '_private_property' not in m_repr

